firebase-firestore/src/main/java/com/google/firebase/firestore/bundle/BundleReader.java

```diff
@@ -0,0 +1,198 @@
+   */
+  public BundleElement getNextElement() throws IOException, JSONException {
+    // Makes sure metadata is read before proceeding. The metadata element is the first element
+    // in the bundle stream.
+    getBundleMetadata();
+    return readNextElement();
+  }
+
+  /** Returns the number of bytes processed so far. */
+  public long getBytesRead() {
+    return bytesRead;
+  }
+
+  public void close() throws IOException {
+    dataReader.close();
+  }
+
+  /**
+   * Reads from the head of internal buffer, Pulls more data from underlying stream until a complete
+   * element is found (including the prefixed length and the JSON string).
+   *
+   * <p>Once a complete element is read, it is dropped from internal buffer.
+   *
+   * <p>Returns either the bundled element, or null if we have reached the end of the stream.
+   */
+  @Nullable
+  private BundleElement readNextElement() throws IOException, JSONException {
+    int length = readLength();
+    if (length == -1) {
+      return null;
+    }
+
+    String json = readJsonString(length);
+    bytesRead += (int) (Math.log10(length) + 1) + length;
+    return BundleElement.fromJson(serializer, json);
+  }
+
+  /**
+   * Reads the length prefix from the beginning of the internal buffer until the first '{'. Returns
+   * the integer-decoded length.
+   *
+   * <p>If it reached the end of the stream, returns -1.
+   */
+  private int readLength() throws IOException {
+    int nextOpenBracket;
+
+    while ((nextOpenBracket = indexOfOpenBracket()) == -1) {
+      if (!pullMoreData()) {
+        break;
+      }
+    }
+
+    // We broke out of the loop because underlying stream is closed, and there happens to be no
+    // more data to process.
+    if (buffer.remaining() == 0) {
+      return -1;
+    }
+
+    // We broke out of the loop because underlying stream is closed, but still cannot find an
+    // open bracket.
+    if (nextOpenBracket == -1) {
+      raiseError("Reached the end of bundle when a length string is expected.");
+    }
+
+    char[] c = new char[nextOpenBracket];
+    buffer.get(c);
+    return Integer.parseInt(new String(c));
+  }
+
+  /** Returns the index of the first open bracket, or -1 if none is found. */
+  private int indexOfOpenBracket() {
+    buffer.mark();
+    try {
+      for (int i = 0; i < buffer.limit(); ++i) {
+        if (buffer.get() == '{') {
+          return i;
+        }
+      }
+      return -1;
+    } finally {
+      buffer.reset();
+    }
+  }
+
+  /**
+   * Reads from a specified position from the internal buffer, for a specified number of bytes,
+   * pulling more data from the underlying stream if needed.
+   *
+   * <p>Returns a string decoded from the read bytes.
+   */
+  private String readJsonString(int length) throws IOException {
+    char[] c = new char[length];
+
+    int read = Math.min(length, buffer.remaining());
+    buffer.get(c, 0, read);
+
+    while (read < length) {
+      if (!pullMoreData()) {
```

Comment:
This is a wasted copy. There's no reason to copy into the buffer only to use that to copy into the character array. You might as well just directly read into `c`. However, creating a `String` from a character array is yet another copy...

Probably the right resolution here is to make `c` a `StringBuilder` and then `append` the buffer contents into it. `CharBuffer` implements `CharSequence` so this should be straightforward.

Note that filling the buffer the way you're doing isn't the most efficient way to do this. For as many bytes as you've read, you might as well append them directly into the target StringBuilder.

Reply:
I tried to figure out how to use a StringBuilder here without making an additional copy. I didn't realize that I could just use `StringBuilder.append()`. Updated the code.

Thanks for the help!